kernel:  x86: MMU: kconfig definition for Memory Management Unit

AdithyaBaglody opened the pull request:
kconfig definition needed to enable the Memory Management Unit in x86.

JIRA: ZEP-2093

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
kernel: MMU: kconfig definition to enable the MMU on x86.

Kconfig definition for enabling the memory management Unit
on x86 based platforms.

JIRA: ZEP-2093

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
kernel: x86: MMU: Macros & Linker scripts for Boot time table creation

Macro is used to create a structure to specify the boot time
page table configuration. Needed by the gen_mmu.py script to generate
the actual page tables.

Linker script is needed for the following:
     1. To place the MMU page tables at 4KByte boundary.
     2. To keep the configuration structure created by
        the Macro(mentioned above).

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
kernel: x86: MMU: Build system support for x86 MMU

Makefile rule to create the MMU page tables at boot time. This
rule invokes the gen_mmu.py script to create a binary which is
then placed into the kernel image using objcopy.
Makefile.mmu is included only when CONFIG_X86_MMU is enabled.

JIRA: ZEP-2095

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
qemu: MMU: Memory regions for qemu_x86

Define the memory regions needed to generate boot time page
tables.

JIRA:ZEP-2098

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
scripts: x86: MMU: Script to generate page tables

Generates the MMU page tables. This creates a binary output for
the page tables. It takes an binary input which represents the
configuration information needed to generate the page tables.

JIRA:ZEP-2096

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

andrewboie commented on arch/x86/Makefile.mmu:
```diff
@@ -0,0 +1,29 @@
+	GENMMU_EXTRA_ARGS :=
+endif
+
+ifeq ($(PREBUILT_HOST_TOOLS),)
```
You don't need this. Just point it to scripts/gen_mmu.py.
This is only for C-based build tools.

andrewboie commented on arch/x86/soc/ia32/soc.c:
```diff
@@ -17,3 +17,16 @@
+#ifdef CONFIG_MMU
+
+#ifdef CONFIG_XIP
+MMU_BOOT_REGION(CONFIG_PHYS_LOAD_ADDR, CONFIG_ROM_SIZE*1024, 1234);
```
what does 1234 parameter mean?
Need to fix the flags values for all these calls.
Should be in terms of mmustructs.h macros.

AdithyaBaglody replied:
This was to test the script in the initial stages. Will change this

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,425 @@
+"""
+template_disabled_pde = """entry[PDE_NUMBER].pt = {.value = 0 }"""
+
+template_enabled_pde = """    .entry[PDE_NUMBER].pt = {
```
I see value in having human-readable debugging output. But if I remember right, we can't generate the PDE in C because we can't actually populate the .page_table field with a valid memory address since it has to be shifted 12 bits.
Instead of outputting C, maybe just print out a human-readable table of the page directory / table contents to stdout? I think this is better than outputting C, someone might see the generated C code and try to build it.

AdithyaBaglody replied:
sure Andrew I will look into creating a simple table which goes to stdout on verbose output flag

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,425 @@
+                                               pte_valid_addr_end = pte_valid_addr_end,
+                                               permissions = basic_mem_region_values[2])
+
+            # print(mem_region_values)
```
get rid of these commented out print statments.
or better yet, create a debug() function which only prints them out in Verbose mode, see debug() in gen_isr_tables.py

AdithyaBaglody replied:
I think I will remove these. I had kept this for my debugging only. the information relevant to the user will be populated elsewhere.

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,425 @@
+    global read_buff
+    global page_tables_list
+    global PD_start_addr
+    global validation_issue_memory_overlap
```
get rid of this variable and just throw an exception as soon as a problem is detected

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,425 @@
+        # print(hex(mem_region_values.start_addr),hex(mem_region_values.pde_index),
+        #       hex(mem_region_values.size))
+        if(mem_region_values.pde_index in list_of_pde.keys()):
+            validation_issue_memory_overlap = True
```
here you should throw an exception, with details on what regions are overlapping each other. i think this is better than raising an error in validate_pde_regions() because down there we won't know specifically which regions overlap each other.

AdithyaBaglody replied:
I didn't want to do too many things in that function so I separated them into 2 different functions.
I found a way to inform the user that a overlap has occurred in the memory with the memory address and Page directory entry number.

andrewboie replied:
at the end of the day, what i'd like to see is if this problem occurs, the user has enough information to immediately know which instances of MMU_BOOT_REGION() are the culprit.

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,425 @@
+                extra_pte_valid_addr_end =  1024
+
+            if(extra_pde in list_of_pde.keys()):
+                validation_issue_memory_overlap = True
```
should throw an exception here

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,425 @@
+                validation_issue_memory_overlap = True
+
+
+                list_of_pde[extra_pde] = mmu_region_details(start_addr = extra_pde_start_address,
```
what is this data used for, given that we know at this point that the memory regions are in an invalid state?

AdithyaBaglody replied:
it look like an indentation issue. Its not supposed to be inside the if condition.

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,425 @@
+
+            extra_pte_valid_addr_end = get_pte_number(mem_region_values.start_addr
+                                                      + overflow_size)
+            # if extra_pte_valid_addr_end >= 1023:
```
remove

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,425 @@
+        for extra_pde in range(pde_index+1, get_pde_number(
+                mem_region_values.start_addr + basic_mem_region_values[1])+1):
+
+            extra_pde_start_address = extra_pde*(4096*1024)
```
There's a lot of math going on here, for all the transformations you do in this function, leave a brief comment explaining why the step is taken.

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,425 @@
+            #     extra_pte_valid_addr_end =  1023
+
+            extra_region_size = overflow_size
+            if overflow_size > (1024*4096):
```
add comment explaining what this is for

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,425 @@
+                present = 1 << 0;
+                read_write = ( ( value.permissions >> 1) & 0x1) << 1;
+                user_mode = ( ( value.permissions >> 2) & 0x1) << 2;
+                pwt = 0 << 3;
```
are you sure these are all correct?
0 shifted by any value is still just 0

AdithyaBaglody replied:
kept it as a place holder.
Whenever there is a need to set that bit , I just have to replace "0" with "1".
Later on I need not worry about bit positions.

andrewboie replied:
ok, fine to leave this, it definitely makes it clear what we are doing.
However, since we are using the #defines in mmustructs,h to parameterize the memory regions, which are already at the bit positions the CPU expects, would it be simpler to just logical OR the flags field in to the table entry?

andrewboie commented on include/arch/x86/linker.ld:
```diff
@@ -137,6 +137,19 @@ SECTIONS
+       /*
+        * Page Tables are located here if MMU is enabled.
+	*/
+	SECTION_DATA_PROLOGUE(mmu_page_tables, (OPTIONAL),)
```
Does this really need to be in its own binary section or can we place this in the data section? For example we put the generated IDT directly in rodata.

AdithyaBaglody replied:
I thought it would make it more understandable/clean than placing the entire thing in .data section. It might look a bit cluttered at the end.

andrewboie replied:
yeah let's not do this unless placing this in its own section solves a specific problem

AdithyaBaglody commented on include/arch/x86/linker.ld:
```diff
@@ -216,6 +229,23 @@ SECTIONS
+	LONG(__mmu_tables_start)
+	__MMU_LIST_START__ = .;
+	KEEP(*(.mmulist))
+	KEEP(*(.gnu.linkonce.mmulist))
```
Since this is a draft version I kept some of these things. 
Sure I will remove them, I meant to remove it in the final version.

andrewboie commented on arch/x86/Kconfig:
```diff
@@ -70,6 +70,14 @@ config X86_IAMCU
 	assembly code will need to be updated to conform to the new calling
 	convention.  If in doubt say N
 
+config MMU
```
rename this to X86_MMU

CONFIG_MMU should be a generic define not specific to any architecture, we will get to it later once the arch-to-kernel code API takes shape.

AdithyaBaglody committed:
kernel: x86: MMU: Enable MMU at boot time.

In crt0.S the MMU is initialized. It uses the statically build
page tables. Here 32-bit paging scheme is used, thereby each page
table entry maps to a 4KB page. The valid regions of the memory are
specified by SOC specific file(soc.c).

JIRA: ZEP-2099

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
quark_se: MMU: Memory regions for arduino 101(quark_se)

Define the memory regions needed to generate boot time page tables.

JIRA: ZEP-2097

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
quark_se: MMU: Add linker support for MMU page table generation.

Add a separate section in the linker to place the MMU configuration
information. This location is read by the gen_mmu.py script to
create the actual page tables.

JIRA: ZEP-2095

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

andrewboie commented on arch/x86/Makefile.mmu:
```diff
@@ -0,0 +1,25 @@
+
+mmu_tables.o: $(PREBUILT_KERNEL)
+	$(call cmd,gen_mmu)
+	$(Q)$(OBJCOPY) -I binary -B $(OUTPUT_ARCH) -O $(OUTPUT_FORMAT) \
```
I think you can move this objcopy call into the cmd_gen_mmu rule

AdithyaBaglody replied:
Andrew I think it is better placed here. Else I have to do some Makefile manipulation to assign the object file to mmu_tables.o from the objcopy output, similar to how it is handled in Makefile.idt.
I feel this is simple and terse.

andrewboie replied:
Huh? All you need to do is transplant that line:

quiet_cmd_gen_mmu = GMMU    $@  
        cmd_gen_mmu =								\  
  (										\  
  	$(OBJCOPY) -I $(OUTPUT_FORMAT)  -O binary -j mmulist $< mmulist.bin &&	\  
  	python $(GENMMU) -i mmulist.bin -o mmu_tables.bin $(GENMMU_EXTRA_ARGS) && \  
        $(Q)$(OBJCOPY) -I binary -B $(OUTPUT_ARCH) -O $(OUTPUT_FORMAT) \ 
                --rename-section .data=.mmu_data mmu_tables.bin $@ \
  )  


It does not make sense to have a 3 command rule split up the way you have it

andrewboie commented on arch/x86/soc/ia32/soc.c:
```diff
@@ -17,3 +17,18 @@
+MMU_BOOT_REGION(CONFIG_PHYS_RAM_ADDR, CONFIG_RAM_SIZE*1024, MMU_ENTRY_WRITE);
+#else
+MMU_BOOT_REGION(CONFIG_PHYS_LOAD_ADDR, CONFIG_RAM_SIZE*1024, MMU_ENTRY_WRITE);
+MMU_BOOT_REGION(CONFIG_LOAPIC_BASE_ADDRESS, 4096*1024, MMU_ENTRY_WRITE);
```
You need to move these APIC definitions outside of the CONFIG_XIP block, they apply in either case.
You also need a memory range for the HPET timer driver if I'm not mistaken.

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,538 @@
+
+
+#consider little endian
+def read_little_endian_binay(file_name):
```
"binary" not "binay"
Actually this function seems unused? You should be using struct module for this

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -283,6 +284,25 @@ bssWords:
+
+	/* load the page directory address into the registers*/
+	movl $__mmu_tables_start, %eax
+	movl %eax,%cr3
```
nitpick: space after comma

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -283,6 +284,25 @@ bssWords:
+	movl %eax, %cr4
+
+	/* set CR0.PG bit (31st bit in CR0)*/
+	movl %cr0,%eax
```
nitpick: space after comma

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -283,6 +284,25 @@ bssWords:
+
+	/* set CR0.PG bit (31st bit in CR0)*/
+	movl %cr0,%eax
+	orl $0x80010000, %eax
```
are there any defined constants in kernel_arch_data.h we could use or add instead of having a magic number directly here?

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -283,6 +284,25 @@ bssWords:
+
+	/*Set CR4.PAE = 0 (5th bit in CR4*/
+	movl %cr4, %eax
+	andl $~0x10, %eax
```
are there any defined constants in kernel_arch_data.h we could use or add instead of having a magic number directly here?

andrewboie commented on arch/x86/soc/intel_quark/quark_se/Kconfig.defconfig.series:
```diff
@@ -17,13 +17,15 @@ config TOOLCHAIN_VARIANT
 
 config PHYS_RAM_ADDR
-	default 0xA8006400
+	default 0xA8007000 if X86_MMU
```
The defaults chosen here were arbitrary. Let's just use A8007000 unconditionally.
Note that you have to also update the quark_se_c1000_ss under arch/arc/soc as well.

AdithyaBaglody replied:
We can continue to use that old value for situations which don't need page alignment. RAM space is never enough. So I think this is better to have this switch needed for MMU.
And on the ARC subsystem, since all my mmu code will be on x86 I don't have anything to modify correct.

andrewboie replied:
The 80K RAM on the SOC is shared between the ARC and x86 sides. So if you change where the X86 RAM starts you need to make a corresponding change on the ARC side configuration.

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,538 @@
+
+    # location from where the Page tables will be written
+    PT_start_addr = pd_start_addr + 4096
+#    print(hex(pd_start_addr))
```
get rid of this

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,538 @@
+            valid_pte = 0
+            for i in value.page_entries_info:
+                temp_value = ((pte >= i.pte_valid_addr_start) and (pte <= i.pte_valid_addr_end))
+            #    print("temp_value",temp_value, pte,i.pte_valid_addr_start, i.pte_valid_addr_end )
```
get rid of this

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,538 @@
+                if temp_value:
+                    perm_for_pte = i.permissions
+                valid_pte |= temp_value
+            # print(value.pte_valid_addr_start,value.pte_valid_addr_end )
```
please remove all these commented out print statements or turn them into debugs

tyrantlizard77 reviewed (changes requested):
I see a large amount of console spam when I try to run this building 'hello_world'.

Please remove whatever is spamming the console. Please also remove all commented out print statements, or turn them into verbose printouts.

andrewboie commented on arch/x86/Makefile.mmu:
```diff
@@ -0,0 +1,25 @@
+)
+
+
+mmu_tables.o: $(PREBUILT_KERNEL)
```
this rule also needs to depend on $(GENMMU) itself in case the script is changed.

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,538 @@
+import os
```
Need to make sure you are using Python 3.
Please add: "#!/usr/bin/env python3"
as the first line of this file.

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,538 @@
+                        help="Input file from which MMU regions are read.")
+    parser.add_argument("-o", "--output",
+                        help="Output file into which the page tables are written.")
+    parser.add_argument("-l", "--list_mmu_table", action="store_true",
```
I would really rather you added this option as "-v" "--verbose".

andrewboie commented on scripts/gen_mmu.py:
```diff
@@ -0,0 +1,538 @@
+          (custom)+ " | "+\
+          page_table_addr +"\n"
+    )
+    print(pde)
```
I'm pretty sure this is what is spamming the console.
This print statement, if it is useful, should provide more surrounding context and not be printed out by default unless the user wants verbose output.

AdithyaBaglody replied:
I needed this for script debugging. I will remove this.

andrewboie commented on arch/x86/soc/ia32/soc.c:
```diff
@@ -17,3 +17,22 @@
+#endif /*CONFIG_XIP */
+
+MMU_BOOT_REGION(CONFIG_LOAPIC_BASE_ADDRESS, 4096*1024, MMU_ENTRY_WRITE);
+MMU_BOOT_REGION(CONFIG_IOAPIC_BASE_ADDRESS, 1024*1024, MMU_ENTRY_WRITE);
```
Are you sure IOAPIC has 1MB address range?
That doesn't correspond to my understanding.
See https://pdos.csail.mit.edu/6.828/2006/readings/ia32/ioapic.pdf
IIRC there is just IOREGSEL and IOWIN registers and thats it.

AdithyaBaglody replied:
I used the one mentioned for quark c1000 as reference
[http://www.intel.com/content/dam/www/public/us/en/documents/datasheets/quark-c1000-datasheet.pdf](http://www.intel.com/content/dam/www/public/us/en/documents/datasheets/quark-c1000-datasheet.pdf)
In this they mention that IOAPIC is 1MB

andrewboie commented on arch/x86/soc/intel_quark/quark_se/soc.c:
```diff
@@ -21,6 +21,34 @@
+MMU_BOOT_REGION(CONFIG_PHYS_RAM_ADDR, CONFIG_RAM_SIZE*1024, MMU_ENTRY_WRITE);
+
+/* loapic */
+MMU_BOOT_REGION(CONFIG_LOAPIC_BASE_ADDRESS, 4*1024, MMU_ENTRY_WRITE);
```
As noted in QEMU's patch I would like you to revisit the size values here.

AdithyaBaglody replied:
I took these values from the datasheet for quark c1000.
[http://www.intel.com/content/dam/www/public/us/en/documents/datasheets/quark-c1000-datasheet.pdf](http://www.intel.com/content/dam/www/public/us/en/documents/datasheets/quark-c1000-datasheet.pdf)

AdithyaBaglody commented on arch/x86/soc/ia32/soc.c:
```diff
@@ -17,3 +17,22 @@
+MMU_BOOT_REGION(CONFIG_PHYS_LOAD_ADDR, CONFIG_RAM_SIZE*1024, MMU_ENTRY_WRITE);
+#endif /*CONFIG_XIP */
+
+MMU_BOOT_REGION(CONFIG_LOAPIC_BASE_ADDRESS, 4096*1024, MMU_ENTRY_WRITE);
```
While debugging I had kept this region to include ioapic as well hence I wrote a big number. I have changed it back to 4KB.

AdithyaBaglody commented on arch/x86/soc/ia32/soc.c:
```diff
@@ -17,3 +17,22 @@
+MMU_BOOT_REGION(CONFIG_IOAPIC_BASE_ADDRESS, 1024*1024, MMU_ENTRY_WRITE);
+
+/*HPET */
+MMU_BOOT_REGION(CONFIG_HPET_TIMER_BASE_ADDRESS, 1024*1024, MMU_ENTRY_WRITE);
```
I couldn't find the start address of this region in that manual. Hence I thought that there maybe a different memory map. I will change it to 4KB.

tyrantlizard77 reviewed (approved):
Thanks, I think we can merge this now.
